src/Servers/Kestrel/Core/src/Internal/Http2/Http2OutputProducer.cs

```diff
@@ -157,19 +249,52 @@ public ValueTask<FlushResult> FlushAsync(CancellationToken cancellationToken)
 
             if (_startedWritingDataFrames)
             {
+                var enqueue = Enqueue(_pipeWriter.UnflushedBytes);
+
                 // If there's already been response data written to the stream, just wait for that. Any header
                 // should be in front of the data frames in the connection pipe. Trailers could change things.
-                return _flusher.FlushAsync(this, cancellationToken);
+                var task = _flusher.FlushAsync(this, cancellationToken);
+
+                if (enqueue)
+                {
+                    Schedule();
+                }
+
+                return task;
             }
             else
             {
-                // Flushing the connection pipe ensures headers already in the pipe are flushed even if no data
-                // frames have been written.
-                return _frameWriter.FlushAsync(this, cancellationToken);
+                var enqueue = EnqueueForObservation(State.FlushHeaders);
+
+                if (enqueue)
+                {
+                    Schedule();
+                }
+
+                return default;
             }
         }
     }
 
+    public void MarkWaitingForWindowUpdates(bool waitingForUpdates)
+    {
+        lock (_dataWriterLock)
+        {
+            _waitingForWindowUpdates = waitingForUpdates;
+        }
+    }
+
+    private void Schedule()
+    {
+        lock (_dataWriterLock)
+        {
+            // Lock here
+            _waitingForWindowUpdates = false;
```

Comment:
Do we do this today?

Reply:
I think `_isScheduled` is sufficient. I doubt the other case where we don't need to schedule because of lack of window are as common. If they are, we can optimize later.